### PR TITLE
add support for gcc

### LINKED ISSUE #11: Support GCC
### Search before asking

- [x] I searched the [issues](https://github.com/alibaba/async_simple/issues) and found no similar issues.

### What happened + What you expected to happen
async_simple only support clang now, we also need to support gcc.

### Anything else
#25 

### Are you willing to submit a PR?

- [ ] Yes I am willing to submit a PR!

### PR DESCRIPTION
**Pull request by foreverhy**
## Why

Close #11 

add gcc support.

### CODE REVIEW

File: async_simple/coro/Task.h [hunk 1]
```diff
@@ -139,8 +139,8 @@ struct TaskPromiseBase {
     }
     void unhandled_exception() const noexcept {}
 
-    void* operator new(size_t sz) noexcept { return ::operator new(sz); }
-    void operator delete(void* p, size_t sz) noexcept {
+    void* operator new(size_t sz) { return ::operator new(sz); }
+    void operator delete(void* p, size_t sz) {
```
Comment (ChuanqiXu9) [inline async_simple/coro/Task.h:143]: Why do we need to remove the `noexcept` specifier?

File: async_simple/coro/test/LazyTest.cpp [hunk 2]
```diff
@@ -420,7 +420,7 @@ TEST_F(LazyTest, testCollectAll) {
 TEST_F(LazyTest, testCollectAllBatched) {
     size_t task_num = 5000;
     long total = 0;
-    for (auto i = 0; i < task_num; i++) {
+    for (int i = 0; i < static_cast<int>(task_num); i++) {
```
Comment (ChuanqiXu9) [inline async_simple/coro/test/LazyTest.cpp:423]: Is this change necessary?

**Comment by ChuanqiXu9** [comment]
It looks like this patch need to be formatted by `git-clang-format HEAD^`.

File: async_simple/coro/Collect.h [hunk 3]
```diff
@@ -94,7 +94,7 @@ struct CollectAnyAwaiter {
         size_t i = 0;
```
Comment (qicosmos) [inline async_simple/coro/Collect.h:94]: ```size_t i = 0``` in for-loop would be better?

File: async_simple/coro/test/LazyTest.cpp [hunk 4]
```diff
@@ -420,7 +420,7 @@ TEST_F(LazyTest, testCollectAll) {
 TEST_F(LazyTest, testCollectAllBatched) {
     size_t task_num = 5000;
     long total = 0;
```
Comment (qicosmos) [inline async_simple/coro/test/LazyTest.cpp:422]: ```long total``` --> ```size_t total``` may be better.
Maybe unified all int to size_t can get ride of static_cast.

File: async_simple/coro/Task.h (hunk 1 above)
Comment (foreverhy) [reply async_simple/coro/Task.h]: Not compile in gcc, Task::promise_type do not have a get_return_object_on_allocation_failure().

File: async_simple/coro/test/LazyTest.cpp (hunk 4 above)
Comment (foreverhy) [reply async_simple/coro/test/LazyTest.cpp:422]: make task_num from `size_t` to `int`.

File: async_simple/coro/Task.h (hunk 1 above)
Comment (ChuanqiXu9) [reply async_simple/coro/Task.h]: It would be better to add `get_return_object_on_allocation_failure()` methods. It could return a Lazy with an exception shows it fails on allocation failure.

File: async_simple/coro/test/LazyTest.cpp (hunk 2 above)
Comment (ChuanqiXu9) [reply async_simple/coro/test/LazyTest.cpp]: ping?

File: async_simple/coro/Task.h (hunk 1 above)
Comment (foreverhy) [reply async_simple/coro/Task.h]: Task do not have an exception_ptr member, which is different from Lazy.

File: async_simple/coro/test/LazyTest.cpp (hunk 2 above)
Comment (foreverhy) [reply async_simple/coro/test/LazyTest.cpp]: now task_num from `size_t` to  `int`, static_cast is unnecessary.

File: async_simple/coro/Task.h (hunk 1 above)
Comment (ChuanqiXu9) [reply async_simple/coro/Task.h]: How about adding `exception_ptr` for Task? Would it be too hard?

File: async_simple/coro/test/LazyTest.cpp [hunk 5]
```diff
@@ -448,7 +448,7 @@ TEST_F(LazyTest, testCollectAllBatched) {
         EXPECT_EQ(task_num, out.size());
         co_await CurrentExecutor();
         int sum = 0;
-        for (auto i = 0; i < out.size(); i++) {
+        for (size_t i = 0; i < out.size(); i++) {
```
Comment (ChuanqiXu9) [inline async_simple/coro/test/LazyTest.cpp:451]: Is this necessary? Does GCC warn if we use `auto` here?

File: async_simple/coro/Task.h (hunk 1 above)
Comment (foreverhy) [reply async_simple/coro/Task.h]: `Task` was created for in-memory applications, maybe we could remove it since nobody is using it.

File: async_simple/coro/test/LazyTest.cpp (hunk 5 above)
Comment (foreverhy) [reply async_simple/coro/test/LazyTest.cpp:451]: `auto` is deduced to `int` here.

**Review by qicosmos — APPROVED** [review]
LGTM

File: async_simple/coro/Task.h (hunk 1 above)
Comment (ChuanqiXu9) [reply async_simple/coro/Task.h]: If nobody uses it, I agree we should remove it.

File: async_simple/coro/test/LazyTest.cpp (hunk 5 above)
Comment (ChuanqiXu9) [reply async_simple/coro/test/LazyTest.cpp:451]: Got it. We'd better use range-based for here. But we could do it in next patches.

File: async_simple/coro/Task.h (hunk 1 above)
Comment (foreverhy) [reply async_simple/coro/Task.h]: Will fix in #29

Comment (ChuanqiXu9) [reply async_simple/coro/Task.h]: OK
